airbyte-integrations/bases/base-typing-deduping-test/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/RecordDiffer.java

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.base.destination.typing_deduping;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.Streams;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.AirbyteProtocolType;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetTime;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * Utility class to generate human-readable diffs between expected and actual records. Assumes 1s1t
+ * output format.
+ */
+public class RecordDiffer {
+
+  private final Comparator<JsonNode> rawRecordIdentityComparator;
+  private final Comparator<JsonNode> rawRecordSortComparator;
+  private final Function<JsonNode, String> rawRecordIdentityExtractor;
+  private final Comparator<JsonNode> finalRecordIdentityComparator;
+  private final Comparator<JsonNode> finalRecordSortComparator;
+  private final Function<JsonNode, String> finalRecordIdentityExtractor;
+
+  /**
+   * @param identifyingColumns Which fields constitute a unique record (typically PK+cursor). Do _not_
+   *        include extracted_at; it is handled automatically.
+   */
+  public RecordDiffer(Pair<String, AirbyteType>... identifyingColumns) {
```

Comment:
this constructor is doing a lot of cool stuff, but maybe too much. It took me a minute to grasp what was happening here. This is a stylistic suggestion, so feel free to ignore, but I feel like the raw and final stuff should be split up. I know that they are supposed to be different but looking at this method the difference is hard to spot. So I think I would do one of a few things:

1. just split between final and raw and let the helper methods set values on the class
2. create methods which build the values which are returned to the constructor, which still sets the values in the class
3. like 2, but reuse logic for both which I think would involve passing a `Function<JsonNode, JsonNode>` which is `r -> r.get("_airbyte_data")` for raw and `r -> r` for final

I think 3 is my preference, its maybe a little weird because java can be annoying about returning multiple things but I think we have `Triplet`s available

Reply:
so the three options are:
1. two `void` methods, one for final, one for raw which set the values in the class, literally just copy and paste out of this constructor
2. similar to one, two different methods but return a `Triplet<Comparator<JsonNode>, Comparator<JsonNode>, Function<JsonNode, String>>` to the constructor and let the constructor set values in the class
3. one method which returns the same triplet type but takes a parameter so you can make changes between raw and final

triplet is just a suggestion for slightly less code rewriting, you could do a further rewrite and have 3 methods, one to build each field type correlating to raw and final

Reply:
or also, do nothing 😄

Reply:
how does https://github.com/airbytehq/airbyte/pull/27852/commits/4442b065118a2543cc70e28a8ec91b5fafebce12 look?

Reply:
that commit looks great